File: 백준/Gold/11082. 소수를 분수로/소수를 분수로.py
# 소수를 분수로
import sys
input = sys.stdin.readline

def getGcd(a, b):
    if b > a:
        a, b = b, a

    while b:
        a, b = b, a % b
        
    return a

num = input().rstrip()
if "." not in num:
    print(num,'/',1, sep='')
    exit()
    
natural, decimal = num.split(".")

if "(" not in decimal:
    numer = int(natural+decimal)
    deno = 10**(len(decimal))
    gcd = getGcd(numer, deno)
    print(numer//gcd, "/", deno//gcd, sep="")
else:
    start = decimal.index("(")
    end = decimal.index(")")

    circulation = decimal[start+1:end]
    deno = '9' * len(circulation)
    deno += '0' * decimal.index("(")
    deno = int(deno)
    
    numer = int(natural + decimal[0:start] + circulation)
    numer -= int(natural+decimal[0:start])

    gcd = getGcd(numer, deno)
    
    print(numer//gcd, "/", deno//gcd, sep="")
    

File: 백준/Gold/11082. 소수를 분수로/소수를 분수로.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String num = br.readLine();

        if(!num.contains(".")) {
            System.out.println(num+"/"+1);
            System.exit(0);
        }

        String[] split = num.split("\\.");
        String natural = split[0];
        String decimal = split[1];

        int gcd;
        if (!decimal.contains("(")) {
            int numer = Integer.parseInt(natural+decimal);
            int deno = (int) Math.pow(10, decimal.length());
            gcd = getGcd(numer, deno);
            System.out.println(numer/gcd +"/"+deno/gcd);
        } else {
            int start = decimal.indexOf("(");
            int end = decimal.indexOf(")");

            String circulation = decimal.substring(start+1, end);
            String denoTmp = "9".repeat(circulation.length());
            denoTmp += "0".repeat(start);
            int deno = Integer.parseInt(denoTmp);

            int numer = Integer.parseInt(natural + decimal.substring(0, start) + circulation);
            numer -= Integer.parseInt(natural + decimal.substring(0, start));

            gcd = getGcd(numer, deno);
            System.out.println(numer/gcd +"/"+deno/gcd);
        }


    }

    static int getGcd(int a, int b) {
        if (b > a) {
            int tmp = a;
            a = b;
            b = tmp;
        }

        while (b != 0) {
            int tmp = b;
            b = a % b;
            a = tmp;
        }

        return a;
    }
}
